src/libp2p_crypto.erl

```diff
@@ -2,119 +2,155 @@
+keys_to_bin(#{secret := {ed25519, PrivKey}, public := {ed25519, PubKey}}) ->
+    io:format("PUB ~p PRIV: ~p", [byte_size(PubKey), byte_size(PrivKey)]),
+    <<?KEYTYPE_ED25519:8, PrivKey:64/binary, PubKey:32/binary>>.
 
--spec from_pem(binary()) -> {ok, private_key(), public_key()}.
-from_pem(PemBin) ->
+-spec keys_from_bin(binary()) -> key_map().
+keys_from_bin(<<?KEYTYPE_ECC_COMPACT:8, PemBin/binary>>) ->
     [PemEntry] = public_key:pem_decode(PemBin),
     PrivKey = public_key:pem_entry_decode(PemEntry),
-    PubKey = make_public_key(PrivKey),
-    {ok, PrivKey, PubKey}.
+    #'ECPrivateKey'{parameters=Params, publicKey=PubPoint} = PrivKey,
+    PubKey = {#'ECPoint'{point=PubPoint}, Params},
+    #{secret => {ecc_compact, PrivKey}, public => {ecc_compact, PubKey}};
+keys_from_bin(<<?KEYTYPE_ED25519, PrivKey:64/binary, PubKey:32/binary>>) ->
+    #{secret => {ed25519, PrivKey}, public => {ed25519, PubKey}}.
 
--spec pubkey_to_address(public_key() | ecc_compact:compact_key()) -> address().
-pubkey_to_address(PubKey) ->
+
+-spec pubkey_to_bin(pubkey()) -> pubkey_bin().
+pubkey_to_bin({ecc_compact, PubKey}) ->
     case ecc_compact:is_compact(PubKey) of
-        {true, CompactKey} -> CompactKey;
+        {true, CompactKey} -> <<?KEYTYPE_ECC_COMPACT, CompactKey/binary>>;
         false -> erlang:error(not_compact)
-    end.
+    end;
+pubkey_to_bin({ed25519, PubKey}) ->
+    <<?KEYTYPE_ED25519, PubKey/binary>>.
 
--spec address_to_pubkey(address()) -> public_key().
-address_to_pubkey(Addr) ->
-    ecc_compact:recover_key(Addr).
+-spec bin_to_pubkey(pubkey_bin()) -> pubkey().
+bin_to_pubkey(<<?KEYTYPE_ECC_COMPACT, PubKey:32/binary>>) ->
+    {ecc_compact, ecc_compact:recover_key(PubKey)};
+bin_to_pubkey(<<?KEYTYPE_ED25519, PubKey:32/binary>>) ->
+    {ed25519, PubKey}.
 
--spec pubkey_to_b58(public_key() | ecc_compact:compact_key()) -> string().
+-spec pubkey_to_b58(pubkey()) -> string().
 pubkey_to_b58(PubKey) ->
-    address_to_b58(pubkey_to_address(PubKey)).
+    bin_to_b58(pubkey_to_bin(PubKey)).
 
--spec b58_to_pubkey(string()) -> public_key().
+-spec b58_to_pubkey(string()) -> pubkey().
 b58_to_pubkey(Str) ->
-    address_to_pubkey(b58_to_address(Str)).
+    bin_to_pubkey(b58_to_bin(Str)).
 
 %% @doc Verifies a digital signature, using sha256.
--spec verify(binary(), binary(), public_key()) -> boolean().
-verify(Bin, Signature, PubKey) ->
-    public_key:verify(Bin, sha256, Signature, PubKey).
+-spec verify(binary(), binary(), pubkey()) -> boolean().
+verify(Bin, Signature, {ecc_compact, PubKey}) ->
+    public_key:verify(Bin, sha256, Signature, PubKey);
+verify(Bin, Signature, {ed25519, PubKey}) ->
+    case enacl:sign_verify_detached(Signature, Bin, PubKey) of
+        {ok, _} -> true;
+        _ -> false
+    end.
+
 
--spec address_to_b58(address()) -> string().
-address_to_b58(Addr) ->
-    address_to_b58(16#00, Addr).
+-spec bin_to_b58(binary()) -> string().
+bin_to_b58(Bin) ->
+    bin_to_b58(16#00, Bin).
```

Comment:
is 16#00 the 'version'?

Reply:
Wondering why the 'version' is not just the key type?

Reply:
Mostly to avoid the types and versions from overlapping.